### PR TITLE
feat: update user data to enable soci on al2023 cpu nodes

### PR DESCRIPTION
**Pull request by debajyoti-truefoundry**

### CODE REVIEW

File: charts/tfy-karpenter-config/README.md [hunk 1]
```diff
@@ -26,15 +27,23 @@ version_lte() {
...
   ! version_lte "$2" "$1"
 }
+
 CONTAINERD_VERSION_OUTPUT="$(containerd --version)"
 IFS=" " read -a  CONTAINERD_VERSION_STRING <<< "${CONTAINERD_VERSION_OUTPUT}"
 CONTAINERD_VERSION=${CONTAINERD_VERSION_STRING[2]}
-echo containerd version "${CONTAINERD_VERSION}"
+SANDBOX_IMAGE=$(containerd config dump | grep -oE 'sandbox_image = "(.*)"' | sed 's/sandbox_image = //')
+KUBELET_CONFIG_DIR=/etc/kubernetes/kubelet/config.json.d
+SOCI_KUBELET_CONFIG_PATH=$KUBELET_CONFIG_DIR/99-soci.conf
```
Comment (chiragjn) [inline charts/tfy-karpenter-config/README.md:36]: Is this always 99-soci.conf?

File: charts/tfy-karpenter-config/README.md [hunk 2]
```diff
@@ -124,15 +126,25 @@ conf_dir = "/etc/cni/net.d"
...
+root = "/var/lib/soci-snapshotter-grpc"
 ##
 EOF
 
+if [ -d "$KUBELET_CONFIG_DIR" ]; then
+cat > $SOCI_KUBELET_CONFIG_PATH << EOF
+  {"apiVersion":"kubelet.config.k8s.io/v1beta1","kind":"KubeletConfiguration","imageServiceEndpoint":"unix:///run/soci-snapshotter-grpc/soci-snapshotter-grpc.sock"}
+EOF
+else
+  echo Kubelet Config dir not found at "$KUBELET_CONFIG_PATH". SOCI will not work for private image.
```
Comment (chiragjn) [inline charts/tfy-karpenter-config/README.md:139]: nit

```suggestion
  echo "Kubelet Config dir not found at $KUBELET_CONFIG_PATH. SOCI will not work for private image."
```

File: charts/tfy-karpenter-config/README.md [hunk 3]
```diff
@@ -3,17 +3,18 @@ A Helm chart for Karpenter provisioners.
...
+3. Adds a configuration file for the `soci-snapshotter-grpc` binary.
 4. Adds a `systemd` service, which manages the `soci-snapshotter-grpc` process.
-5. Patches `containerd` config to use the `soci-snapshotter`.
+5. Patches `containerd` and `kubelet` config to use the `soci-snapshotter`.
 
-### `userData` script for `default` Node Template
+### `userData` script for `default` EC2 Node Class
+* Note that the following script assumes AL2023 and will **not** work on AL2.
 
 ```bash
```
Comment (chiragjn) [inline charts/tfy-karpenter-config/README.md:19]: I was wondering, can we put this in files/
as `al2023-soci-userdata.sh` ?

Easier to maintain and we can add a special value in values.yaml which can inject this userdata on helm render

Comment (debajyoti-truefoundry) [reply charts/tfy-karpenter-config/README.md:19]: Oh yes, I am developing like that only, will add in git.

File: charts/tfy-karpenter-config/README.md (hunk 1 above)
Comment (debajyoti-truefoundry) [reply charts/tfy-karpenter-config/README.md]: You can have multiple configurations that will be merged.

```
[root@ip-10-2-17-231 /]# ls /etc/kubernetes/kubelet/config.json.d
00-nodeadm.conf  99-soci.conf
```

nodeadm adds the default and we patch on top of that.

Comment (debajyoti-truefoundry) [reply charts/tfy-karpenter-config/README.md]: The name does not need to be `99-soci.conf`.

File: charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+yum install fuse -y
+modprobe fuse
+
+curl \
+--silent \
+--show-error \
+--retry 10 \
+--retry-delay 1 \
+-L https://github.com/awslabs/soci-snapshotter/releases/download/v$SOCI_RELEASE_VERSION/soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz \
```
Comment (chiragjn) [inline charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh:36]:
```suggestion
-L "https://github.com/awslabs/soci-snapshotter/releases/download/v${SOCI_RELEASE_VERSION}/soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz" \
```

**Review by chiragjn — APPROVED** [review]

File: charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+-o soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz
```
Comment (debajyoti-truefoundry) [inline charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh:37]: add sha check

**Review by chiragjn — APPROVED** [review]

File: charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+yum install fuse -y
+modprobe fuse
+
+curl \
+--silent \
+--show-error \
+--retry 10 \
+--retry-delay 1 \
+-L https://github.com/awslabs/soci-snapshotter/releases/download/v${SOCI_RELEASE_VERSION}/soci-snapshotter-${SOCI_RELEASE_VERSION}-linux-amd64.tar.gz \
+-o soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz
```
Comment (chiragjn) [inline charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh:37]: Do these curly braces work fine without quoting the argument? I normally quote the string 😅
 

```suggestion
-L "https://github.com/awslabs/soci-snapshotter/releases/download/v${SOCI_RELEASE_VERSION}/soci-snapshotter-${SOCI_RELEASE_VERSION}-linux-amd64.tar.gz" \
-o "soci-snapshotter-${SOCI_RELEASE_VERSION}-linux-amd64.tar.gz"
```

**Review by chiragjn — APPROVED** [review]

**Pull request merged**